File: src/services/goals.ts
/**
 * Goals service + gauntlet orchestrator (§11).
 *
 * The orchestrator is CONTROL-PLANE CODE, not a user-facing agent. After
 * every goal session it reads the progress log + DoD + last session summary,
 * marks satisfied DoD checkboxes, checks the safety rails (spend / time /
 * stuck-at-19), and either completes the goal or spawns the next specialist.
 */
import { randomUUID } from "node:crypto";
import { eq, inArray } from "drizzle-orm";
import { goals, agents, kv } from "../db/schema.js";
import type { DB } from "../db/client.js";
import type { Goal, DoDItem, GoalStatus } from "../domain/types.js";
import { HttpError } from "../api/errors.js";

export interface CreateGoalInput {
  projectId: string;
  title: string;
  spec: string;
  definitionOfDone?: string[];
  spendCapUsd?: number | null;
  maxDurationMinutes?: number | null;
  runnerPreference?: "cloud" | "local" | "auto";
  stuckThreshold?: number;
}

export interface OrchestratorDecision {
  action: "complete" | "continue" | "stop";
  stopReason?: GoalStatus;
  nextAgentId?: string | null;
  nextAgentName?: string | null;
  summary: string;
}

const STUCK_KEY = (goalId: string) => `goal:stuck:${goalId}`;

export class GoalService {
  constructor(private db: DB) {}

  async get(goalId: string): Promise<Goal | null> {
    const row = await this.db.select().from(goals).where(eq(goals.id, goalId)).get();
    if (!row) return null;
    return { ...row, definitionOfDone: row.definitionOfDone ?? [], sessionIds: row.sessionIds ?? [] };
  }

  async list(projectId: string): Promise<Goal[]> {
    const rows = await this.db.select().from(goals).where(eq(goals.projectId, projectId)).all();
    return rows
      .map((r) => ({ ...r, definitionOfDone: r.definitionOfDone ?? [], sessionIds: r.sessionIds ?? [] }))
      .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
  }

  /**
   * §11.2: draft a Definition of Done from the spec sheet when the human did
   * not write one. Heuristic stand-in: bullet items from the spec, else
   * sentence splits. (A planner model call can replace this; labeled.)
   */
  static draftDoD(spec: string): string[] {
    const lines = spec.split("\n").map((l) => l.trim()).filter(Boolean);
    const bullets = lines
      .filter((l) => /^[-*•]|\d+[.)]/.test(l))
      .map((l) => l.replace(/^[-*•]\s*|\d+[.)]\s*/, ""));
    if (bullets.length >= 2) return bullets;
    const sentences = spec
      .split(/(?<=[.!?])\s+/)
      .map((s) => s.trim())
      .filter((s) => s.length > 8);
    return sentences.slice(0, 5);
  }

  async create(input: CreateGoalInput): Promise<Goal> {
    const dodTexts = input.definitionOfDone && input.definitionOfDone.length > 0
      ? input.definitionOfDone
      : GoalService.draftDoD(input.spec);
    if (dodTexts.length === 0) {
      throw new HttpError(400, "could not draft a definition of done — write DoD items explicitly");
    }
    const now = new Date().toISOString();
    const row: Goal = {
      id: randomUUID(),
      projectId: input.projectId,
      title: input.title,
      spec: input.spec,
      definitionOfDone: dodTexts.map((t) => ({ id: randomUUID(), text: t, done: false }) satisfies DoDItem),
      dodApproved: false,
      status: "active",
      spendCapUsd: input.spendCapUsd ?? null,
      spendUsd: 0,
      maxDurationMinutes: input.maxDurationMinutes ?? null,
      stuckThreshold: input.stuckThreshold ?? 19,
      runnerPreference: input.runnerPreference ?? "auto",
      progressLog: "",
      startedAt: null,
      sessionIds: [],
      createdAt: now,
    };
    await this.db.insert(goals).values(row).run();
    return row;
  }

  async approveDoD(goalId: string): Promise<Goal> {
    const g = await this.get(goalId);
    if (!g) throw new HttpError(404, "goal not found");
    if (g.definitionOfDone.every((d) => !d.done) && g.definitionOfDone.length === 0) {
      throw new HttpError(400, "goal has an empty definition of done");
    }
    // §11: a goal without a spend cap is allowed only with explicit human
    // confirmation — the API layer requires `confirmNoCap: true`.
    const updated = {
      ...g,
      dodApproved: true,
      startedAt: g.startedAt ?? new Date().toISOString(),
    };
    await this.db.update(goals).set(updated).where(eq(goals.id, goalId)).run();
    return updated;
  }

  async setStatus(goalId: string, status: GoalStatus): Promise<Goal> {
    const g = await this.get(goalId);
    if (!g) throw new HttpError(404, "goal not found");
    const updated = { ...g, status };
    await this.db.update(goals).set(updated).where(eq(goals.id, goalId)).run();
    return updated;
  }

  /** Append-only progress log. */
  async appendProgress(goalId: string, entry: string): Promise<Goal> {
    const g = await this.get(goalId);
    if (!g) throw new HttpError(404, "goal not found");
    const stamp = `[${new Date().toISOString()}] ${entry}`;
    const progressLog = g.progressLog ? g.progressLog + "\n" + stamp : stamp;
    await this.db.update(goals).set({ progressLog }).where(eq(goals.id, goalId)).run();
    return { ...g, progressLog };
  }

  async addSession(goalId: string, sessionId: string): Promise<void> {
    const g = await this.get(goalId);
    if (!g) return;
    await this.db
      .update(goals)
      .set({ sessionIds: [...g.sessionIds, sessionId] })
      .where(eq(goals.id, goalId))
      .run();
  }

  async addSpend(goalId: string, usd: number): Promise<Goal> {
    const g = await this.get(goalId);
    if (!g) throw new HttpError(404, "goal not found");
    const spendUsd = Math.round((g.spendUsd + usd) * 10000) / 10000;
    await this.db.update(goals).set({ spendUsd }).where(eq(goals.id, goalId)).run();
    return { ...g, spendUsd };
  }

  async setDoDItems(goalId: string, items: DoDItem[]): Promise<Goal> {
    const g = await this.get(goalId);
    if (!g) throw new HttpError(404, "goal not found");
    const updated = { ...g, definitionOfDone: items };
    await this.db.update(goals).set(updated).where(eq(goals.id, goalId)).run();
    return updated;
  }

  /**
   * Mark ONE definition-of-done item satisfied. This is the only way an item
   * becomes done: a specialist calls it after actually finishing and
   * verifying the item — the orchestrator never infers completion from
   * progress-log text. Matching is exact first, then containment, then word
   * overlap; a non-matching call fails instead of guessing.
   */
  async completeDoDItem(goalId: string, itemText: string): Promise<Goal> {
    const g = await this.get(goalId);
    if (!g) throw new HttpError(404, "goal not found");
    const target = normalizeText(itemText);
    if (!target) throw new HttpError(400, "empty definition-of-done item");
    const items = g.definitionOfDone.map((d) => ({ ...d }));

    let idx = items.findIndex((d) => !d.done && normalizeText(d.text) === target);
    if (idx === -1) {
      idx = items.findIndex(
        (d) => !d.done && (normalizeText(d.text).includes(target) || target.includes(normalizeText(d.text))),
      );
    }
    if (idx === -1) {
      let best = -1;
      let bestScore = 0;
      items.forEach((d, i) => {
        if (d.done) return;
        const s = missionScore(d.text, target) + missionScore(target, d.text);
        if (s > bestScore) {
          bestScore = s;
          best = i;
        }
      });
      idx = bestScore > 0 ? best : -1;
    }
    if (idx === -1) {
      throw new HttpError(400, `no unsatisfied definition-of-done item matches "${itemText}"`);
    }
    items[idx] = { ...items[idx]!, done: true };
    return this.setDoDItems(goalId, items);
  }

  // -------------------------------------------------------------------------
  // Orchestrator
  // -------------------------------------------------------------------------

  /** §11.6: run after every goal session. Returns the decision taken. */
  async orchestrate(
    goalId: string,
    opts: { allowList: { id: string; name: string }[] },
  ): Promise<OrchestratorDecision> {
    const g = await this.get(goalId);
    if (!g) throw new HttpError(404, "goal not found");
    if (!g.dodApproved) {
      return { action: "stop", stopReason: "stopped-stuck", summary: "goal not approved; orchestrator refuses to spawn" };
    }
    if (g.status !== "active") return { action: "stop", summary: `goal status ${g.status}` };

    // --- Safety rails (§11) ------------------------------------------------
    if (g.spendCapUsd !== null && g.spendUsd >= g.spendCapUsd) {
      await this.setStatus(goalId, "stopped-spend");
      return { action: "stop", stopReason: "stopped-spend", summary: `spend ${g.spendUsd} >= cap ${g.spendCapUsd}` };
    }
    if (g.maxDurationMinutes !== null && g.startedAt) {
      const elapsed = (Date.now() - new Date(g.startedAt).getTime()) / 60000;
      if (elapsed >= g.maxDurationMinutes) {
        await this.setStatus(goalId, "stopped-time");
        return { action: "stop", stopReason: "stopped-time", summary: `elapsed ${Math.round(elapsed)}m >= ${g.maxDurationMinutes}m` };
      }
    }

    // --- DoD satisfaction (explicit, agent-marked) ---------------------------
    // An item is done only when the specialist marked it via
    // goals.complete_dod_item — never inferred from free-text progress logs.
    const items = g.definitionOfDone;
    const allDone = items.every((d) => d.done);
    if (allDone) {
      await this.setStatus(goalId, "completed");
      return { action: "complete", summary: "all DoD checkboxes satisfied" };
    }

    // --- Stuck detection: same specialist + no progress delta, N times -------
    const last = await this.lastSessionInfo(g);
    const stuckRow = await this.db.select().from(kv).where(eq(kv.key, STUCK_KEY(goalId))).get();
    const stuck = stuckRow ? JSON.parse(stuckRow.value) : { lastAgent: null as string | null, lastProgressLen: 0, count: 0 };
    const progressLen = g.progressLog.length;
    const sameAgent = stuck.lastAgent === last.agentName;
    const noDelta = progressLen === stuck.lastProgressLen;
    let count = 0;
    if (sameAgent && noDelta && last.agentName) count = stuck.count + 1;
    else count = sameAgent ? stuck.count : 0;
    const threshold = g.stuckThreshold;
    await this.db
      .insert(kv)
      .values({ key: STUCK_KEY(goalId), value: JSON.stringify({ lastAgent: last.agentName, lastProgressLen: progressLen, count }) })
      .onConflictDoUpdate({ target: kv.key, set: { value: JSON.stringify({ lastAgent: last.agentName, lastProgressLen: progressLen, count }) } })
      .run();
    if (count >= threshold) {
      await this.setStatus(goalId, "stopped-stuck");
      return { action: "stop", stopReason: "stopped-stuck", summary: `${count} identical iterations without progress` };
    }

    // --- Pick next specialist ------------------------------------------------
    // DoD-aware dispatch: score each specialist's mission against the first
    // unsatisfied DoD item; fall back to round-robin when nothing matches
    // (keeps the loop deterministic and test-stable).
    const firstOpenItem = items.find((d) => !d.done)?.text ?? "";
    const missionByAgent = await this.missionsFor(opts.allowList);
    const next = this.pickNextAgent(opts.allowList, last.agentName, firstOpenItem, missionByAgent);
    if (!next) {
      await this.setStatus(goalId, "stopped-stuck");
      return { action: "stop", stopReason: "stopped-stuck", summary: "no specialist available on the allow list" };
    }
    return { action: "continue", nextAgentId: next.id, nextAgentName: next.name, summary: `spawn ${next.name}` };
  }

  /** Role-prompt mission + deliverable sections per allow-listed agent (dispatch input). */
  private async missionsFor(
    allowList: { id: string; name: string }[],
  ): Promise<Map<string, string>> {
    const map = new Map<string, string>();
    if (allowList.length === 0) return map;
    const rows = await this.db
      .select()
      .from(agents)
      .where(inArray(agents.id, allowList.map((a) => a.id)))
      .all();
    for (const a of rows) {
      map.set(a.name, `${a.name} ${a.title} ${promptSection(a.rolePrompt, "## Mission")} ${promptSection(a.rolePrompt, "## Deliverable")}`);
    }
    return map;
  }

  /**
   * §8.2: choose the next specialist. Prefer the specialist whose profile
   * (name, title, mission, deliverable) best covers the first unsatisfied
   * DoD item's significant words; ties rotate by recency, and no match
   * falls back to the full round-robin so a 2-item DoD spawns ≥2 sessions.
   */
  private pickNextAgent(
    allowList: { id: string; name: string }[],
    lastAgentName: string | null,
    firstOpenItem: string,
    profileByAgent: Map<string, string>,
  ): { id: string; name: string } | null {
    let bestScore = 0;
    const tied: { id: string; name: string }[] = [];
    if (firstOpenItem) {
      for (const a of allowList) {
        const score = missionScore(profileByAgent.get(a.name) ?? "", firstOpenItem);
        if (score > bestScore) {
          bestScore = score;
          tied.length = 0;
          tied.push(a);
        } else if (score > 0 && score === bestScore) {
          tied.push(a);
        }
      }
    }
    if (bestScore > 0) return rotate(tied, lastAgentName);

    // Round-robin through the whole allow list (no keyword match).
    return rotate(allowList, lastAgentName);
  }

  private async lastSessionInfo(g: Goal): Promise<{ agentName: string | null; progressLen: number }> {
    const id = g.sessionIds[g.sessionIds.length - 1];
    if (!id) return { agentName: null, progressLen: 0 };
    const row = await this.db.query.sessions.findFirst({ where: (s, { eq: e }) => e(s.id, id) });
    if (!row) return { agentName: null, progressLen: 0 };
    const agent = row.agentId ? await this.db.select().from(agents).where(eq(agents.id, row.agentId)).get() : null;
    return { agentName: agent?.name ?? null, progressLen: g.progressLog.length };
  }
}

/** The text of one "## Section" of a structured role prompt (dispatch input). */
function promptSection(rolePrompt: string, marker: string): string {
  const i = rolePrompt.indexOf(marker);
  if (i < 0) return "";
  const rest = rolePrompt.slice(i + marker.length);
  const j = rest.indexOf("##");
  return (j >= 0 ? rest.slice(0, j) : rest).trim();
}

/** Lowercase + collapse non-alphanumerics to single spaces. */
function normalizeText(s: string): string {
  return s.toLowerCase().replace(/[^a-z0-9 ]+/g, " ").replace(/\s+/g, " ").trim();
}

/** Rotate a list so the agent that ran last is moved to the back, then take the head. */
function rotate<T extends { name: string }>(list: T[], lastAgentName: string | null): T | null {
  const out = [...list];
  if (lastAgentName) {
    const idx = out.findIndex((a) => a.name === lastAgentName);
    if (idx >= 0) out.push(out.splice(0, idx + 1).shift()!);
  }
  return out.length ? out[0]! : null;
}

const STOPWORDS = new Set([
  "the", "and", "for", "with", "from", "into", "this", "that", "your", "you",
  "are", "not", "when", "will", "then", "each", "only", "can", "its",
  "item", "a", "an", "of", "to", "in", "on", "by", "at", "or", "as", "be",
  "it", "is", "one", "any", "all", "but", "has", "have", "was", "were",
]);

/** Coverage of a DoD item's significant words by a mission text (0..1). */
export function missionScore(mission: string, itemText: string): number {
  const norm = (s: string) =>
    s.toLowerCase().replace(/[^a-z0-9 ]+/g, " ").split(/\s+/).filter(Boolean);
  const words = norm(itemText).filter((w) => w.length > 3 && !STOPWORDS.has(w));
  if (words.length === 0) return 0;
  const wanted = new Set(words);
  const missionWords = norm(mission);
  let hits = 0;
  for (const w of missionWords) if (wanted.has(w)) hits++;
  return hits / words.length;
}

File: agentos/services/goals.py
from __future__ import annotations
"""Goals service + orchestrator. Port of src/services/goals.ts."""
import json
import re
import uuid
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy import select

from agentos.db.client import get_session
from agentos.db.models import Agent as AgentRow, Goal as GoalRow, KV, Session as SessionRow


class HttpError(Exception):
    def __init__(self, status: int, message: str) -> None:
        super().__init__(message)
        self.status = status
        self.message = message


_STOPWORDS = {
    "the", "and", "for", "with", "from", "into", "this", "that", "your", "you",
    "are", "not", "when", "will", "then", "each", "only", "can", "its",
    "item", "a", "an", "of", "to", "in", "on", "by", "at", "or", "as", "be",
    "it", "is", "one", "any", "all", "but", "has", "have", "was", "were",
}


def _norm(s: str) -> list[str]:
    return re.sub(r"[^a-z0-9 ]+", " ", s.lower()).split()


def mission_score(mission: str, item_text: str) -> float:
    words = [w for w in _norm(item_text) if len(w) > 3 and w not in _STOPWORDS]
    if not words:
        return 0.0
    wanted = set(words)
    hits = sum(1 for w in _norm(mission) if w in wanted)
    return hits / len(words)


def _normalize_text(s: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"[^a-z0-9 ]+", " ", s.lower())).strip()


def _rotate(lst: list[dict], last_name: str | None) -> dict | None:
    if not lst:
        return None
    out = list(lst)
    if last_name:
        idx = next((i for i, a in enumerate(out) if a["name"] == last_name), -1)
        if idx >= 0:
            out.append(out.pop(0 if idx == 0 else idx))
    return out[0] if out else None


class GoalService:
    @staticmethod
    def draft_dod(spec: str) -> list[str]:
        lines = [l.strip() for l in spec.split("\n") if l.strip()]
        bullets = [re.sub(r"^[-*•]\s*|\d+[.)]\s*", "", l) for l in lines if re.match(r"^[-*•]|\d+[.)]", l)]
        if len(bullets) >= 2:
            return bullets
        sentences = [s.strip() for s in re.split(r"(?<=[.!?])\s+", spec) if len(s.strip()) > 8]
        return sentences[:5]

    async def get(self, goal_id: str) -> dict | None:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.id == goal_id))
            row = result.scalar_one_or_none()
        return _row_to_dict(row) if row else None

    async def list(self, project_id: str) -> list[dict]:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.project_id == project_id).order_by(GoalRow.created_at.desc()))
            return [_row_to_dict(r) for r in result.scalars().all()]

    async def create(self, input: dict) -> dict:
        dod_texts = input.get("definitionOfDone") or self.draft_dod(input["spec"])
        if not dod_texts:
            raise HttpError(400, "could not draft a definition of done — write DoD items explicitly")
        now = datetime.now(timezone.utc).isoformat()
        row = GoalRow(
            id=str(uuid.uuid4()),
            project_id=input["projectId"],
            title=input["title"],
            spec=input["spec"],
            definition_of_done=[{"id": str(uuid.uuid4()), "text": t, "done": False} for t in dod_texts],
            dod_approved=False,
            status="active",
            spend_cap_usd=input.get("spendCapUsd"),
            spend_usd=0.0,
            max_duration_minutes=input.get("maxDurationMinutes"),
            stuck_threshold=input.get("stuckThreshold", 19),
            runner_preference=input.get("runnerPreference", "auto"),
            progress_log="",
            started_at=None,
            session_ids=[],
            created_at=now,
        )
        async with get_session() as db:
            db.add(row)
            await db.commit()
        return _row_to_dict(row)

    async def approve_dod(self, goal_id: str) -> dict:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.id == goal_id))
            row = result.scalar_one_or_none()
            if not row:
                raise HttpError(404, "goal not found")
            row.dod_approved = True
            row.started_at = row.started_at or datetime.now(timezone.utc).isoformat()
            await db.commit()
        return await self.get(goal_id)  # type: ignore[return-value]

    async def set_status(self, goal_id: str, status: str) -> dict:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.id == goal_id))
            row = result.scalar_one_or_none()
            if not row:
                raise HttpError(404, "goal not found")
            row.status = status
            await db.commit()
        return await self.get(goal_id)  # type: ignore[return-value]

    async def append_progress(self, goal_id: str, entry: str) -> dict:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.id == goal_id))
            row = result.scalar_one_or_none()
            if not row:
                raise HttpError(404, "goal not found")
            stamp = f"[{datetime.now(timezone.utc).isoformat()}] {entry}"
            row.progress_log = (row.progress_log + "\n" + stamp) if row.progress_log else stamp
            await db.commit()
        return await self.get(goal_id)  # type: ignore[return-value]

    async def add_session(self, goal_id: str, session_id: str) -> None:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.id == goal_id))
            row = result.scalar_one_or_none()
            if row:
                row.session_ids = list(row.session_ids or []) + [session_id]
                await db.commit()

    async def add_spend(self, goal_id: str, usd: float) -> dict:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.id == goal_id))
            row = result.scalar_one_or_none()
            if not row:
                raise HttpError(404, "goal not found")
            row.spend_usd = round((row.spend_usd or 0) + usd, 4)
            await db.commit()
        return await self.get(goal_id)  # type: ignore[return-value]

    async def complete_dod_item(self, goal_id: str, item_text: str) -> dict:
        async with get_session() as db:
            result = await db.execute(select(GoalRow).where(GoalRow.id == goal_id))
            row = result.scalar_one_or_none()
            if not row:
                raise HttpError(404, "goal not found")
            items = list(row.definition_of_done or [])
            target = _normalize_text(item_text)
            if not target:
                raise HttpError(400, "empty definition-of-done item")

            idx = next((i for i, d in enumerate(items) if not d["done"] and _normalize_text(d["text"]) == target), -1)
            if idx == -1:
                idx = next((i for i, d in enumerate(items) if not d["done"] and (
                    _normalize_text(d["text"]) in target or target in _normalize_text(d["text"])
                )), -1)
            if idx == -1:
                best, best_score = -1, 0.0
                for i, d in enumerate(items):
                    if d["done"]:
                        continue
                    s = mission_score(d["text"], target) + mission_score(target, d["text"])
                    if s > best_score:
                        best_score, best = s, i
                idx = best if best_score > 0 else -1
            if idx == -1:
                raise HttpError(400, f'no unsatisfied DoD item matches "{item_text}"')

            items[idx] = {**items[idx], "done": True}
            row.definition_of_done = items
            await db.commit()
        return await self.get(goal_id)  # type: ignore[return-value]

    async def orchestrate(self, goal_id: str, allow_list: list[dict]) -> dict:
        g = await self.get(goal_id)
        if not g:
            raise HttpError(404, "goal not found")
        if not g["dodApproved"]:
            return {"action": "stop", "stopReason": "stopped-stuck", "summary": "goal not approved"}
        if g["status"] != "active":
            return {"action": "stop", "summary": f"goal status {g['status']}"}

        # Safety rails
        if g["spendCapUsd"] is not None and g["spendUsd"] >= g["spendCapUsd"]:
            await self.set_status(goal_id, "stopped-spend")
            return {"action": "stop", "stopReason": "stopped-spend",
                    "summary": f"spend {g['spendUsd']} >= cap {g['spendCapUsd']}"}
        if g["maxDurationMinutes"] and g["startedAt"]:
            from datetime import timezone as tz
            elapsed = (datetime.now(tz.utc) - datetime.fromisoformat(g["startedAt"])).total_seconds() / 60
            if elapsed >= g["maxDurationMinutes"]:
                await self.set_status(goal_id, "stopped-time")
                return {"action": "stop", "stopReason": "stopped-time",
                        "summary": f"elapsed {int(elapsed)}m >= {g['maxDurationMinutes']}m"}

        # DoD complete?
        if all(d["done"] for d in g["definitionOfDone"]):
            await self.set_status(goal_id, "completed")
            return {"action": "complete", "summary": "all DoD checkboxes satisfied"}

        # Stuck detection
        last_agent = await self._last_agent_name(g)
        stuck_key = f"goal:stuck:{goal_id}"
        async with get_session() as db:
            result = await db.execute(select(KV).where(KV.key == stuck_key))
            kv_row = result.scalar_one_or_none()
        stuck = json.loads(kv_row.value) if kv_row else {"lastAgent": None, "lastProgressLen": 0, "count": 0}
        progress_len = len(g["progressLog"])
        same_agent = stuck["lastAgent"] == last_agent
        no_delta = progress_len == stuck["lastProgressLen"]
        count = (stuck["count"] + 1) if (same_agent and no_delta and last_agent) else (stuck["count"] if same_agent else 0)
        new_stuck = {"lastAgent": last_agent, "lastProgressLen": progress_len, "count": count}
        async with get_session() as db:
            if kv_row:
                result = await db.execute(select(KV).where(KV.key == stuck_key))
                kv_row2 = result.scalar_one_or_none()
                if kv_row2:
                    kv_row2.value = json.dumps(new_stuck)
            else:
                db.add(KV(key=stuck_key, value=json.dumps(new_stuck)))
            await db.commit()
        if count >= g["stuckThreshold"]:
            await self.set_status(goal_id, "stopped-stuck")
            return {"action": "stop", "stopReason": "stopped-stuck",
                    "summary": f"{count} identical iterations without progress"}

        # Pick next specialist
        first_open = next((d["text"] for d in g["definitionOfDone"] if not d["done"]), "")
        profiles = await self._missions_for(allow_list)
        next_agent = self._pick_next_agent(allow_list, last_agent, first_open, profiles)
        if not next_agent:
            await self.set_status(goal_id, "stopped-stuck")
            return {"action": "stop", "stopReason": "stopped-stuck", "summary": "no specialist available"}
        return {"action": "continue", "nextAgentId": next_agent["id"],
                "nextAgentName": next_agent["name"], "summary": f"spawn {next_agent['name']}"}

    async def _missions_for(self, allow_list: list[dict]) -> dict[str, str]:
        if not allow_list:
            return {}
        ids = [a["id"] for a in allow_list]
        async with get_session() as db:
            result = await db.execute(select(AgentRow).where(AgentRow.id.in_(ids)))
            rows = result.scalars().all()
        return {
            r.name: f"{r.name} {r.title} {_prompt_section(r.role_prompt, '## Mission')} {_prompt_section(r.role_prompt, '## Deliverable')}"
            for r in rows
        }

    def _pick_next_agent(self, allow_list: list[dict], last_name: str | None,
                          first_open: str, profiles: dict[str, str]) -> dict | None:
        best_score, tied = 0.0, []
        if first_open:
            for a in allow_list:
                score = mission_score(profiles.get(a["name"], ""), first_open)
                if score > best_score:
                    best_score, tied = score, [a]
                elif score > 0 and score == best_score:
                    tied.append(a)
        if best_score > 0:
            return _rotate(tied, last_name)
        return _rotate(allow_list, last_name)

    async def _last_agent_name(self, g: dict) -> str | None:
        session_ids = g.get("sessionIds", [])
        if not session_ids:
            return None
        last_id = session_ids[-1]
        async with get_session() as db:
            result = await db.execute(select(SessionRow).where(SessionRow.id == last_id))
            s = result.scalar_one_or_none()
            if not s:
                return None
            result2 = await db.execute(select(AgentRow).where(AgentRow.id == s.agent_id))
            agent = result2.scalar_one_or_none()
        return agent.name if agent else None


def _prompt_section(role_prompt: str, marker: str) -> str:
    i = role_prompt.find(marker)
    if i < 0:
        return ""
    rest = role_prompt[i + len(marker):]
    j = rest.find("##")
    return (rest[:j] if j >= 0 else rest).strip()


def _row_to_dict(r: GoalRow) -> dict:
    return {
        "id": r.id, "projectId": r.project_id, "title": r.title, "spec": r.spec,
        "definitionOfDone": r.definition_of_done or [], "dodApproved": r.dod_approved,
        "status": r.status, "spendCapUsd": r.spend_cap_usd, "spendUsd": r.spend_usd,
        "maxDurationMinutes": r.max_duration_minutes, "stuckThreshold": r.stuck_threshold,
        "runnerPreference": r.runner_preference, "progressLog": r.progress_log,
        "startedAt": r.started_at, "sessionIds": r.session_ids or [], "createdAt": r.created_at,
    }
